src/app-logic/url-handling.js

```diff
@@ -48,9 +48,23 @@ let _isReplaceState: boolean = false;
  * history API's behavior.
  */
 export function setHistoryReplaceState(value: boolean): void {
+  if (_isReplaceState === value) {
+    throw new Error(
+      `Trying to toggle the history replace state to: ${value.toString()} but it was already ${_isReplaceState.toString()}.`
+    );
+  }
   _isReplaceState = value;
 }
 
+/**
+ * This function only handles sync callbacks, but async could be handled if needed.
+ */
+export function withHistoryReplaceState(fn: () => void): void {
+  setHistoryReplaceState(true);
+  fn();
+  setHistoryReplaceState(false);
```

Comment:
It occurred to me that we may want to check the initial value when starting, rather than setting to `false` unconditionnally at the end. This is to help with cases where this would be used in a nested way:
```js
withHistoryReplaceState(() => {
  withHistoryReplaceState(() => {
    // do something
  });
});
```
This looks silly written like this, but I believe this can happen unknowingly when dispatching a thunk action from another thunk action.

Here is my suggestion:
```js
export function withHistoryReplaceState(fn: () => void): void {
  if (getIsHistoryReplaceState()) {
    fn();
  } else {
    setHistoryReplaceState(true);
    fn();
    setHistoryReplaceState(false);
  }
};
```
I think it's better to use this `if` condition rather than saving the value at the start of the function, because we don't want to set the value to `true` (which would happen in a nested case) if it was set to `false` for some reason during the function execution.

Reply:
I think my preference here would be to either throw an error if it was already in replace state, or go all-in and handle all the cases. I'm going flagging you for review again with the "complete" solution. I was avoiding it initially because of the complexity of implementation.

Reply:
Actually, it already throws an error in this case, so it can't be used as suggested. See the implementation of `setHistoryReplaceState`.

```js
  if (_isReplaceState === value) {
    throw new Error(
      `Trying to toggle the history replace state to: ${value.toString()} but it was already ${_isReplaceState.toString()}.`
    );
  }
  _isReplaceState = value;
```

Reply:
Ok, I'm going to keep the current implementation, as I believe it's correct, and break this out into a follow-up PR.